src/Servers/Kestrel/Core/src/Internal/Http2/AspNetCoreHPackEncoder.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http.HPack;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
+{
+    internal class AspNetCoreHPackEncoder
+    {
+        private HeaderEnumerator _enumerator;
+
+        private HPackEncoder _hPackEncoder = new HPackEncoder();
+
+        public bool BeginEncode(HeaderEnumerator enumerator, Span<byte> buffer, out int length)
```

Comment:
This logic is copied and pasted from `HPackEncoder` and changed to use `HeaderEnumerator` instead of `IEnumerable`. This type can't live in shared code because the shared code won't know about `HeaderEnumerator`.

If the client doesn't use BeginEncode/Encode then those methods could be removed from shared code.

Reply:
No, I don't think HttpClient uses BeginEncode. It looks like they use primarily static methods.
https://github.com/dotnet/runtime/blob/e9853d4baa4c9510dc62ed5852f8381141f3c87e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L948
Try removing it in the runtime repo first.